root/static/scripts/edit/URLCleanup.js

```diff
@@ -7201,17 +7480,82 @@ const CLEANUPS: CleanupEntries = {
 };
 /* eslint-enable sort-keys */
 
-function testAll(tests: $ReadOnlyArray<RegExp>, text: string) {
-  for (let i = 0; i < tests.length; i++) {
-    if (tests[i].test(text)) {
-      return true;
+function findCleanupEntryByHostname(
+  hostname: string,
+  sourceUrl: string,
+): CleanupEntry | null {
+  const entries = CLEANUP_ENTRIES_BY_HOSTNAME[hostname];
+  if (entries) {
+    for (let i = 0; i < entries.length; i++) {
+      const entry = entries[i];
+      const matchRegExps = entry.match;
+      for (let i = 0; i < matchRegExps.length; i++) {
+        if (matchRegExps[i].test(sourceUrl)) {
+          return entry;
+        }
+      }
     }
   }
+  return null;
+}
 
-  return false;
+const httpRegExp = /^https?:\/\//;
+function findCleanupEntry(inputUrl: string): CleanupEntry | null {
+  const url = httpRegExp.test(inputUrl) ? inputUrl : ('http://' + inputUrl);
+  const {
+    domain,
+    domainWithoutSuffix,
+    hostname,
+    subdomain,
+  } = tldtsParse(url);
+  if (!domain) {
+    return null;
+  }
+  let entry = (
+    findCleanupEntryByHostname(domain, url) ||
+    findCleanupEntryByHostname(hostname, url) ||
+    findCleanupEntryByHostname(domainWithoutSuffix + '.*', url)
+  );
+  if (entry) {
+    return entry;
+  }
+  if (subdomain) {
+    const subdomainLabels = subdomain.split('.');
+    let subdomainPart = '';
+    for (let i = subdomainLabels.length - 1; i >= 0; i--) {
+      subdomainPart = subdomainLabels[i] + '.' + subdomainPart;
+      entry = (
+        findCleanupEntryByHostname(subdomainPart + domain, url) ||
+        findCleanupEntryByHostname(
+          subdomainPart + domainWithoutSuffix + '.*',
+          url,
+        )
+      );
+      if (entry) {
+        return entry;
+      }
+    }
+  }
+  return null;
 }
 
-const CLEANUP_ENTRIES: Array<CleanupEntry> = Object.values(CLEANUPS);
+export const CLEANUP_ENTRIES_BY_HOSTNAME:
```

Comment:
Isn't this almost a rewrite of `groupBy` in `arrays.js` with an object instead of a map? It seems to work anyway and this is a big improvement, so I'll merge and release, but just wondering if we can simplify it further.

Reply:
It's similar, but `groupBy` only supports returning a single key from the callback, so placing each value in a single group. In this case, an entry can be assigned to multiple groups (since `hostname` can be an array).

Reply:
Ok, I guess we don't want to expand `groupBy` to allow for this, so then this makes sense.